### ISSUE TITLE
Does not work on Mountain Lion

### ISSUE THREAD

**Issue by e1z0** [ayurmedia/xnu-speedstep-air#2]
Hi,

I've compiled the module for mountain lion and it seems that it does not work and i think it needs to be compiled specifically with 10.8 sdk instead of 10.9. Dmesg shows this when trying to load kext on mountain lion:

 
![IMG_5206](https://user-images.githubusercontent.com/7213361/155884237-d902c2b3-774c-44b8-8dd9-35ae071fbe25.jpeg)

But then trying to compile it using 10.8 sdk a lot of errors shows up.

![IMG_5205](https://user-images.githubusercontent.com/7213361/155884084-aed0f593-36df-4f94-80b7-5810fdbb72e5.jpeg)

Any advice is appreciated, thanks in advance!

**Comment by ayurmedia** [comment]
Hi, i haven't used this kext in ages. gone through several versions of macs and osx. 

xcode projects are a little bit compatible to other versions than what you build it for. in IOS for iPad you can still run old Apps in the new OSX. but you cannot run new apps in the old OSX if they are too different. Apple often changes the API, which roughly translates to access points which functions you can all, in the end the "kernel" is also just software but with special behaviour. 

Ideally your target is the same version as your OSX which the kext will run in. I think i fixed small things compared to the original xnu-speedstep (link is at the end of the readme) to get it run on newer osx versions. But at some point the kext does not work at all as OSX changes the API. Coolbook sometimes works. But mostly in old Macs. 

The "depreciation" warnings in Xcode you can just ignore, it just means "in the future" this might not work. 
the red errors "Semantic issue" is important and stops the compilation. and indicates the code and the SDK dont work together. We are now many many versions away from 10.8 so this is all good luck and experimental. 
but at the time i published the project on github it was working on my computer, means the code is ok and compiles. but the environment changed and xcode etc. means its good luck to get it working now. 

Anyway OSX 10.8 will be quite useless nowadays as most apps wont work, ssl will be broken and you cannot use internet in a useful way. basically you can only run the software which comes on he CD if you are offline. 
If you still have this old hardware it might be better to install linux on it and make it a little bit useful and practical. 

also when you try to install new OSX on old hardware. after 3 years difference things will start to break and not work as expected. apple wants you to buy the latest hardware and osx and does not care about computers older than 3 years. you can manage to extend the lifetime of macs but you get into problems eventually if you are to far away from the mainstream. 

you should also google for coolbook, it might work for your mac. i am not the original author of this kext, i just upgraded the code to be compatible with the version which was newest at the time and the old didnt compile anymore. 
in new osx (my current laptop) its getting worse and worse and there is no control over speedstep. the only trick in new systems to prevent overheating is "volta" which can limit the power consumption and in a result slow down the whole system. but its not like undervolting, because with volta you actually loose performance, but it might help to cool the system. 

good luck

**Comment by e1z0** [comment]
I'm just old vintage collector who wants to make old macbook pro 2,1 2006 17" run mountain lion smoothly, the only problem is here are overheating it reaches 100C very fast, the thermal paste didn't helped either, so i tried coolbook which is currently not supported on lion or mountain lion, i've tried to remove Apple power management kexts and try to run it but i get error #14 and it wont run ant all.

**Comment by ayurmedia** [comment]
you still might get coolbook to run with some tricks. you have to google it, i don't remember how exactly but i managed to get it running in newer versions than the official supported by coolbook. its quite tricky. you have to mess around with the kexts , sometimes this "sleepenabler.kext" involved etc. different generation of cpus. modern macs don't have any of that. and i can understand old macs (even new) have a general problem of overheating if you do a little bit more than read your emails :) especially the old macs don't have the hardware, so even watching 720p youtube is a heavy task. 
But as you figured out the way too high voltages in the speedstep from official osx lead to overheating. reducing it a little helps quite a lot. for me replacing heatpaste on the CPU did help. but it's more the os. Did you check the logs (istat menus) if maybe there is some crazy process running using all CPU instead of idle? For me having scaled resolution on mirrored external monitor leads to overheat. sometimes you find solutions. i recommend "volta" app and "istats menu" to figure out what is going on.i wonder if volta works on this
 old osx. 

<img width="342" alt="Screenshot 2022-02-27 at 16 36 25" src="https://user-images.githubusercontent.com/1746990/155889035-41b82062-8360-4972-aace-3441fcdaad9d.png">
 
Volta 

<img width="318" alt="Screenshot 2022-02-27 at 16 37 53" src="https://user-images.githubusercontent.com/1746990/155889085-9276ec60-87a9-4343-a881-2cc8dcd48854.png">

istat menu. 
you can read a lot of detective information from the power consumption and figure out what is running at the time and effects. like you can even see the power draw difference changing brightness of the laptop screen. 

also in old macs replacing the hdd (spinning disk) with an SSD (Sata) makes a huge difference in speed. they are very cheap now, and with a time maschine backup its easy to copy the os to the new drive. (just make backup on old, and restore it to new. might have to have a basic install on the new drive first and then restore the backup). 

does not really solve the problem, but gets the fan noise more controllable and constant and you can trade of heat vs. noise.

<img width="828" alt="Screenshot 2022-02-27 at 16 42 28" src="https://user-images.githubusercontent.com/1746990/155889229-8fb594db-7caa-4937-88cd-696e020c6410.png">
macs fan control (i think this app works from 10.5 to 10.11 :) 

Have fun, maybe you manage to compile this xnu-speedstep. treat it as a hobby, because old osx are broken and not all is fixable. 

https://forums.macrumors.com/threads/coolbook-working-on-mavericks-details-here.1662099/
You might still get lucky with coolbook and some tricks. i once got it working on one more osx generation outside the official release.

**Comment by e1z0** [comment]
I have tried to tune fans to the maximum, i still reach up to 105C in this macbook already have istatmenus, which i'm using to monitor voltages, temp and other info. Also managed to run coolbook binary /usr/bin/CoolBookSwitcher and kext sucessfully, but i need manually edit `se.coolbook.switcher.plist` to do any tuning because the GUI app CoolBookController exits with error 21. Does not know what it means at all. I will try to run it from terminal to see if it outputs some debug info if it does not i will try to dissasemble and fix it in hard way with hex editor. :)

**Comment by ayurmedia** [comment]
Overheating over 100C is not normal. Either some App/Process is running wild. Eg. Spotlight Indexer can run for hours, or backup etc. Or the Hardware has some problems. Thermal-Paste, Heatsync. Fan, Dust etc. One thing not to underestimate is climate. It's always funny, when friends complain in summer that the fans run so much, and then in winter they magically "fixed it" :) If the heat cannot escape in the air, the fan does not help. But i tried myself, hence optimizing this kext. over 90C there is thermal throttling and it slows down and gets less responsive. OSX kind of has a bug there, then it starts some kernel process to slow the processor, but in endeffect it still heats up. once the temp is lower this kernel task disappears. This means if the CPU is throttling of overheat. in summary it might be faster to lower the frequency of the cpu and its still faster. also this turbo cpu peaks are not really good for temperature, on my 2013 macbook its quite inefficient strategy how cpu and fans run. but its optimized for sprints and then idle. high cpu for fast action and then low cpu while you read or do other stuff. macs are not made for number crunching and converting video etc. the laptops are optimized for light usage. the "pro" is a joke in new ones. But the new M1 and M1X change the game, i will transition when they get cheaper. 

i have uploaded an old kext i found in some old folders. the Info.plist inside says built for "osx 10.7" , maybe you are lucky and it works for you. 10.7 vs. 10.8 might be compatible. newer projects seem to start with 10.9 as SDK in xcode. (check the plists inside, there are different prebuilt binaries)

https://github.com/ayurmedia/xnu-speedstep-air/tree/master/Tools
IntelEnhancedSpeedStep_10.7.kext.zip

seems the info.plist in that kext also has a list of the voltages and frequencies. but as i remember one needs some tricks to disable the normal speedstep or osx will overrule the xnu-speedstep a second later. but reading via command line still works. 
in the Tools folder are some scripts to debug and play around and check if it works. 

Did you try the prebuilt kexts in the Tools folder ?
https://github.com/ayurmedia/xnu-speedstep-air/blob/master/Tools/IntelEnhancedSpeedStepLow.kext/Contents/Info.plist

They are built for 10.8, the version you are using. i even put some .command scripts there where you can load and unload the kext and watch the output of the cpu values. 
in .kext/Contents/Info.plist you can see the Target Versions. seems there are some for 10.7, 10.8 and 10.9. You can just edit the plist with normal editor to change the frequencies and voltages without recompiling. 
The StepLow.kext and StepHigh.kext in the Tools folder only differ in the manually edited Plist (config) the "binary" is the same. 

Good Luck. Seems 2014 i was playing around with it in osx 10.8 and 10.9. But eventually used Coolbook with the trick or switched back to the automatic stepping of OSX as the differenced where not as expected. Coolbook actually made a big difference in older versions. This undervolting works. Some years ago i even undervolted my then windows laptop and managed to run it without fans totally silent. (maybe not the best for the lifetime of the CPU :) On Apple, there is also some lower limits to undervolt and it gets unstable and crashes if the CPU is starved of electrons. normally the voltages are dynamic to the usage and calculations....

also, but you probably tried is to make sure air-flow is good with a computer stand which lifts the laptop, its problem with all macs, maybe the new M1X is improving some as some new macs are fanless. 

Good Luck, i close this ticket as i cannot help more. 

If you find some solutions you can make a pull request and i will merge it. but i think the audience of this project is quite small :)

**Comment by ayurmedia** [comment]
Just for fun i tried to compile the project again, but on xcode-13 the oldest SDK is 10.9 . So you cannot build it towards 10.8. and there are some changes in the kernel and it cannot be changed and upgraded easily. many of the APIs dont exist in osx 11.6 etc. its impossible to load those kexts. 

your biggest chance is actually to install xcode on the target macbook like Xcode 4.x and compile it there, as you cannot cross-compile from a new osx 11.x its just too far away. 

Or go the easy route and just use the prebuilt kexts in the Tools folder of this Github project and manually edit the Plist files which are plaintext and read on runtime without compiling. they should be 10.8 and solve your problem. you can find the frequencies lists on wikipedia etc. for your model. i think i was using macbook2,1 at the time. 

But according your screenshots this is on the target 10.8 already. guessing from the user-interface and colorful close buttons and the way istats menu looks like. 
then i dont understand why you not just set the target to 10.8 in SDK and compile again. the project from 2014 was optimized for 10.8. I am sure it was running on my old 2008 Macbook Unibody. So the code itself works. but building with wrong SDK probably does not work. also the Kext must match the OSX i think. maybe compatible in one direction. But in some newer versions the API changed, thats the main reason also Coolbook broke. By chance some depreciated APIs still work on Generation later. Eg. XNU-Speedstep still worked, while Coolbook was broke, as they use a bit different ways to call the APIs. Also i don't really understand that CPP code, i just made some changes to get it to compile. The guy making Coolbook probably actually knows how it works, but decided to make it commercial and abandon it once the API changes / broke. 

Must simple is to just download the code as ZIP from https://github.com/ayurmedia/xnu-speedstep-air and use the kext and scripts (.command = .sh) and play around. you can open the kext with "shop Package contents" or via command line it shows up as folders. you need to unzip the _10.7.kext.zip (Github compressed it when i uploaded it). 
also the .command you can simply doubleclick like apps in osx finder. many have a wait or sleep so the terminal stays open. also you might need to change permission to executable (chmod 777 or similar) to get the scripts to run. but i assume you know that already when you tinker with these things on a low level. 

TLDR; use the prebuild kext and scripts in the Tools folder for 10.8, no need to compile it. You can still edit the .plists to change the frequencies without compiling as the plist is read as a config in runtime and not compile time. 

you can load/unload kexts either by putting it into the system/library (see readme) and reboot. or more simple with kextload and kextunload dynamically (i think you need to be root, or use sudo). the kext either loads and works. or it shows a error message that the kext could not be loaded. The readme should explain what is needed to get it running. I think you need to move away AppleIntelCPUPowerManagement.kext, or else the build in speedstep is resetting the cpu values of the new kext all the time and nothing happens.

**Comment by e1z0** [comment]
Thank you for your time and detailed respond, appreciate it. Will test it soon, thanks a lot :)

**Comment by e1z0** [comment]
Tried to load your provided IntelEnhancedSpeedStep_10.7.kext.zip but was unable to load

![IMG_5207](https://user-images.githubusercontent.com/7213361/155926888-d7ab86a9-ce9c-4abc-9586-5436e1293dbc.jpeg)
